### PR TITLE
draft: per-dso .so resolution cache on glibc

### PR DESCRIPTION
**Pull request by pennae**
###### Description of changes

this is an alternative to #207061. to reduce stat storms and general slowness during application startup we save a resolution cache into each dso that maps DT_NEEDED entries to their corresponding path in the configured runpath. this should Just Work™.

we think this has many benefits over that, shrinkwrap, and other solutions:
- unlike shrinkwrap, dependencies can still be replaced with `LD_LIBRARY_PATH`
- unlike the guix approach it's in theory possible for a soname to resolve to two different paths in the same process (if that's possible at all in glibc, haven't actually checked. should be possible with multiple namespaces though)
- no new files in outputs are needed that could conflict when merging
- works when cross-compiling without qemu-user involvement

we've tested this on a cross-compiled armv7 system (in qemu-user, got bored of the board taking so long to load) running `strace -cf /run/current-system/bin/switch-to-configuration test`. unpatched nixos takes 44s to do this, with these patches that drops to 29s owing to almost 24000 failing `openat` and `stat` calls that are avoided by using the cache.

this is still very rough, full of debug code, tests are missing, and the patchelf code should be upstreamed (but is included here for ease of testing).

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by blaggacao** [comment]
I've read through all the linked materials and wonder what's the fundamental critique of an approach that leverages nix' stable store locator properties to evict a system that originally was made to accommodate _unstable_ locators with magic?

The comparison to the other approaches doesn't seem to sufficiently convey that insatisfaction, in my view.

Especially, since they seem to approach the problem from first principles.

**Comment by fzakaria** [comment]
Interesting approach - might be worth sharing the ideas with the GNU libc or musl mailing list.
Maybe there's an idea here worth expanding to the linking specification.

Can you show using `readelf -x .secion-name` to demonstrate the output.

here's a real question:
If we are willing to patch the dynamic loader (ldso) -- why even put all this information into the single executable and have to muck with patchelf?

This is the question I keep asking. We need to challenge the notion that we must even have a single binary.

**Comment by blaggacao** [comment]
Ofc, Farid just wanted to avoid another shameless plug, so here it is for what he means: https://github.com/NixOS/nixpkgs/pull/207061#issuecomment-1360787824

😁

File: pkgs/development/libraries/glibc/common.nix [hunk 1]
```diff
@@ -167,6 +169,7 @@ stdenv.mkDerivation ({
   # out as the first output is an exception exclusive to glibc
   outputs = [ "out" "bin" "dev" "static" ];
 
+  dontStrip = true;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/glibc/common.nix:172]: We should comment why this is required

File: pkgs/development/libraries/glibc/ldcache.patch [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
++      _dl_debug_printf("    found notes %lx size %lx\n", (long) nhdr, (long) size);
++    while (size >= sizeof *nhdr) {
++      if (__glibc_unlikely (GLRO(dl_debug_mask) & DL_DEBUG_LIBS))
++        _dl_debug_printf("      check note namesz %x descsz %x\n", nhdr->n_namesz, nhdr->n_descsz);
++      const size_t noteSize = sizeof *nhdr + ALIGN_UP(nhdr->n_namesz, 4) + ALIGN_UP(nhdr->n_descsz, 4);
++      if (size < noteSize) break;
++      size -= noteSize;
++      const char *nname = (const char *) (nhdr + 1);
++      const char *desc = (const char *) (nhdr + 1) + ALIGN_UP(nhdr->n_namesz, 4);
++      if (nhdr->n_namesz == sizeof ELF_NOTE_NIXOS
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/glibc/ldcache.patch:31]: at least a few comments would not hurt

**Comment by edolstra** [comment]
Can you describe in a bit more detail what the resolution process at runtime looks like? In particular, what happens when `LD_LIBRARY_PATH` is set.

In principle this approach seems good to me, though I'm slightly worried about maintaining a non-trivial patch to glibc. I wonder if upstream glibc might be interested in this...

File: pkgs/development/tools/misc/patchelf/ldcache.patch [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
++
++    for (const auto &[dep, dir] : cache)
++        debug("resolved %s to %s\n", dep.c_str(), dir.c_str());
++
++    const auto noteSize = std::accumulate(
++        cache.begin(), cache.end(), 1, // extra trailing '\0'
++        [](auto acc, const auto & p) { return acc + p.first.size() + p.second.size() + 2; });
++
++    static const char NOTE_NAME[] = ".note.nixos.ldcache";
++    const auto nameSize = roundUp(strlen(ELF_NOTE_NIXOS) + 1, 4);
```
Comment (hinshun) [inline pkgs/development/tools/misc/patchelf/ldcache.patch:152]:
```suggestion
+    const auto nameSize = roundUp(strlen(ELF_NOTE_NIXOS) + 1, getPageSize());
```

Comment (hinshun) [inline pkgs/development/tools/misc/patchelf/ldcache.patch:152]:
```suggestion
+    const auto nameSize = roundUp(strlen(ELF_NOTE_NIXOS) + 1, getPageSize());
```

File: pkgs/development/tools/misc/patchelf/ldcache.patch [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
++    const auto descSize = roundUp(noteSize, 4);
```
Comment (hinshun) [inline pkgs/development/tools/misc/patchelf/ldcache.patch:153]:
```suggestion
+    const auto descSize = roundUp(noteSize, getPageSize());
```

File: pkgs/development/tools/misc/patchelf/ldcache.patch [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
++        wri(loadPhdr.p_align, getPageSize());
++        wri(hdr()->e_phnum, rdi(hdr()->e_phnum) + 1);
++
++        auto & notePhdr = phdrs.emplace_back();
++        wri(notePhdr.p_type, PT_NOTE);
++        wri(notePhdr.p_offset, noteOffset);
++        wri(notePhdr.p_vaddr, wri(notePhdr.p_paddr, noteAddr));
++        wri(notePhdr.p_filesz, wri(notePhdr.p_memsz, cacheSectionSize));
++        wri(notePhdr.p_flags, PF_R);
++        wri(notePhdr.p_align, 4);
```
Comment (hinshun) [inline pkgs/development/tools/misc/patchelf/ldcache.patch:226]:
```suggestion
+        wri(notePhdr.p_align, getPageSize());
```

File: pkgs/development/tools/misc/patchelf/ldcache.patch [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
++
++        // this addition will let us replaceSection(.shstrtab), and thus rewriteSections()!
++        auto & noteShdr = shdrs.emplace_back();
++        wri(noteShdr.sh_name, sectionNames.size());
++        wri(noteShdr.sh_type, SHT_NOTE);
++        wri(noteShdr.sh_flags, SHF_ALLOC);
++        wri(noteShdr.sh_addr, noteAddr);
++        wri(noteShdr.sh_offset, noteOffset);
++        wri(noteShdr.sh_size, cacheSectionSize);
++        wri(noteShdr.sh_addralign, 4);
```
Comment (hinshun) [inline pkgs/development/tools/misc/patchelf/ldcache.patch:237]:
```suggestion
+        wri(noteShdr.sh_addralign, sectionAlignment);
```

**Review by hinshun — COMMENTED** [review]
We're been experimenting with this patch internally with positive results. Compared to the other solutions, this one has a few advantages:
- [Shrinkwrapping](https://github.com/NixOS/patchelf/pull/357) or [setting absolute sonames](https://github.com/fzakaria/nix-harden-needed) causes issues in a local development workflow (outside of nix) where the same shared object from two different nix closures will both be loaded
- LD_LIBRARY_PATH and LD_PRELOAD continue to function, this is very important for us in existing local / debugging workflows

Without the suggestions in this review, I'm seeing page/section alignment errors from `ld`.

File: pkgs/development/tools/misc/patchelf/ldcache.patch [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
++    // TODO tests for:
++    //   - add note to {static,dynamic} {library,executable}
++    //   - update note
++    //   - set-interpreter with note in place (initramfs extra-utils)
++    //   - set-rpath with note in place (initramfs extra-utils)
++    for (; rdi(dyn->d_tag) != DT_NULL; dyn++) {
++        if (rdi(dyn->d_tag) == DT_NEEDED) {
++            needed.emplace_back(strTab + rdi(dyn->d_un.d_val));
++        } else if (rdi(dyn->d_tag) == DT_RUNPATH) {
++            runpath = splitColonDelimitedString(strTab + rdi(dyn->d_un.d_val));
```
[the remark this answers is not in the document]
Comment (hinshun) [reply pkgs/development/tools/misc/patchelf/ldcache.patch:118]: Can we not just ignore the variables and continue processing the other colon delimited paths?

**Comment by mweinelt** [comment]
@DavHau Could you make sure to properly credit @pennae as required by the MIT license?

**Comment by DavHau** [comment]
Please forgive me. Credited now via commit message.

**Pull request closed without merging**